File: azure-ts-webserver-component/webserver.ts
import * as pulumi from "@pulumi/pulumi";
import * as azure from "@pulumi/azure";
import { specialArchiveSig } from "@pulumi/pulumi/runtime";

// All resources will share a resource group.
let resourceGroupName = new azure.core.ResourceGroup("server").name;

// Create a network and subnet for all VMs.
let network = new azure.network.VirtualNetwork("server-network", {
    resourceGroupName,
    addressSpaces: [ "10.0.0.0/16" ],
    subnets: [{
        name: "default",
        addressPrefix: "10.0.1.0/24",
    }],
});
let subnet = new azure.network.Subnet("server-subnet", {
    resourceGroupName,
    virtualNetworkName: network.name,
    addressPrefix: "10.0.2.0/24",
});

/**
 * WebServer is a reusable web server component that creates and exports a NIC, public IP, and VM.
 */
export class WebServer extends pulumi.ComponentResource {
    public readonly publicIp: azure.network.PublicIp;
    public readonly networkInterface: azure.network.NetworkInterface;
    public readonly vm: azure.compute.VirtualMachine;

    /**
     * Allocate a new web server VM, NIC, and public IP address.
     * @param name The name of the web server resource.
     * @param args A bag of arguments to control the web server VM creation.
     */
    constructor(name: string, args: WebServerArgs) {
        super("ws-ts-azure-comp:webserver:WebServer", name);

        // Allocate a public IP and assign it to our NIC.
        this.publicIp = new azure.network.PublicIp(`${name}-ip`, {
            resourceGroupName,
            allocationMethod: "Dynamic",
        }, { parent: this });
        this.networkInterface = new azure.network.NetworkInterface(`${name}-nic`, {
            resourceGroupName,
            ipConfigurations: [{
                name: "webserveripcfg",
                subnetId: subnet.id,
                privateIpAddressAllocation: "Dynamic",
                publicIpAddressId: this.publicIp.id,
            }],
        }, { parent: this });

        // Now create the VM, using the resource group and NIC allocated above.
        this.vm = new azure.compute.VirtualMachine(`${name}-vm`, {
            resourceGroupName,
            networkInterfaceIds: [ this.networkInterface.id ],
            vmSize: args.vmSize || "Standard_A0",
            deleteDataDisksOnTermination: true,
            deleteOsDiskOnTermination: true,
            osProfile: {
                computerName: "hostname",
                adminUsername: args.username,
                adminPassword: args.password,
                customData: args.bootScript,
            },
            osProfileLinuxConfig: {
                disablePasswordAuthentication: false,
            },
            storageOsDisk: {
                createOption: "FromImage",
                name: `${name}-osdisk1`,
            },
            storageImageReference: {
                publisher: "canonical",
                offer: "UbuntuServer",
                sku: "16.04-LTS",
                version: "latest",
            },
        }, { parent: this });
    }

    public getIpAddress(): pulumi.Output<string> {
        // The public IP address is not allocated until the VM is running, so wait for that
        // resource to create, and then lookup the IP address again to report its public IP.
        let ready = pulumi.all({
            _: this.vm.id, name: this.publicIp.name, resourceGroupName: this.publicIp.resourceGroupName });
        return ready.apply(d =>
            azure.network.getPublicIP({
                name: d.name, resourceGroupName: d.resourceGroupName }).then(ip => ip.ipAddress));
    }
}

export interface WebServerArgs {
    /**
     * A required username for the VM login.
     */
    username: pulumi.Input<string>;
    /**
     * A required encrypted password for the VM password.
     */
    password: pulumi.Input<string>;
    /**
     * An optional boot script that the VM will use.
     */
    bootScript?: pulumi.Input<string>;
    /**
     * An optional VM size; if unspecified, Standard_A0 (micro) will be used.
     */
    vmSize?: pulumi.Input<string>;
}

File: azure-py-webserver-component/webserver.py
from pulumi import Config, export, asset, Output, ComponentResource, ResourceOptions
from pulumi_azure import core, compute, network


class WebServerArgs:
    def __init__(
        self,
        resource_group: core.ResourceGroup,
        subnet: network.Subnet,
        username: str,
        password: str,
    ):
        self.resource_group = resource_group
        self.subnet = subnet
        self.username = username
        self.password = password


class WebServer(ComponentResource):
    def __init__(self, name: str, args: WebServerArgs, opts: ResourceOptions = None):
        super().__init__("custom:app:WebServer", name, {}, opts)

        child_opts = ResourceOptions(parent=self)

        public_ip = network.PublicIp(
            "server-ip",
            resource_group_name=args.resource_group.name,
            location=args.resource_group.location,
            allocation_method="Dynamic",
            opts=child_opts,
        )

        network_iface = network.NetworkInterface(
            "server-nic",
            resource_group_name=args.resource_group.name,
            location=args.resource_group.location,
            ip_configurations=[
                {
                    "name": "webserveripcfg",
                    "subnet_id": args.subnet.id,
                    "private_ip_address_allocation": "Dynamic",
                    "public_ip_address_id": public_ip.id,
                }
            ],
            opts=child_opts,
        )

        userdata = """#!/bin/bash
        echo "Hello, World!" > index.html
        nohup python -m SimpleHTTPServer 80 &"""

        vm = compute.VirtualMachine(
            "server-vm",
            resource_group_name=args.resource_group.name,
            location=args.resource_group.location,
            network_interface_ids=[network_iface.id],
            vm_size="Standard_A0",
            delete_data_disks_on_termination=True,
            delete_os_disk_on_termination=True,
            os_profile={
                "computer_name": "hostname",
                "admin_username": args.username,
                "admin_password": args.password,
                "custom_data": userdata,
            },
            os_profile_linux_config={"disable_password_authentication": False},
            storage_os_disk={"create_option": "FromImage", "name": "myosdisk1"},
            storage_image_reference={
                "publisher": "canonical",
                "offer": "UbuntuServer",
                "sku": "16.04-LTS",
                "version": "latest",
            },
            opts=child_opts,
        )

        # The public IP address is not allocated until the VM is running, so we wait
        # for that resource to create, and then lookup the IP address again to report
        # its public IP.
        combined_output = Output.all(
            vm.id, public_ip.name, public_ip.resource_group_name
        )
        self.public_ip_addr = combined_output.apply(
            lambda lst: network.get_public_ip(name=lst[1], resource_group_name=lst[2]).ip_address
        )
        self.register_outputs({})
